README.md
weibo-mid [![Build Status](https://secure.travis-ci.org/fengmk2/weibo-mid.png)](http://travis-ci.org/fengmk2/weibo-mid) [![Coverage Status](https://coveralls.io/repos/fengmk2/weibo-mid/badge.png)](https://coveralls.io/r/fengmk2/weibo-mid)
=======

![logo](https://raw.github.com/fengmk2/weibo-mid/master/logo.png)

Convert [Weibo Open API](http://open.weibo.com) mid to base62 hash string.

## Install

```bash
$ npm install weibo-mid
```

## Usage

```js
var mid = require('weibo-mid');

mid.encode('3600375418559878').should.equal('A09LvzUOy');
mid.decode('A09LvzUOy').should.equal('3600375418559878');
```

## License 

(The MIT License)

Copyright (c) 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.1 / 2013-07-16 
==================

  * add component support
  * 100% coverage
  * add logo
  * init, add mid.encode() and mid.decode()


test/mid.test.js
/*!
 * weibo-mid - test/mid.test.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

try {
  var mid = require('../');
  var should = require('should');
} catch (e) {
  // component
  var mid = require('weibo-mid');
  var should = require('chai').should();
}

describe('mid.test.js', function () {
  var expects = [
    ['1', '1'],
    ['0', '0'],
    ['111111', 'sU7'],
    ['11010123', '14eMj'],
    ['1024', 'gw'],
    ['1024012131', '1EgPJN'],
    ['3600375418559878', 'A09LvzUOy'],
    ['201110410216293360', 'wr4mOFqpbO'],
    ['231101124784859058', 'Bh0tgako3U'],
    ['3491273850170657', 'yCirT0Iox'],
  ];

  it('should encode success', function () {
    expects.forEach(function (item) {
      mid.encode(item[0]).should.equal(item[1]);
    });

    mid.encode('foo').should.equal('foo');
  });

  it('should decode success', function () {
    expects.forEach(function (item) {
      mid.decode(item[1]).should.equal(item[0]);
    });
  });
});


lib/mid.js
/*!
 * weibo-mid - lib/mid.js
 * 
 * 新浪微博mid与url互转实用工具
 * 初始版本作者: @XiNGRZ (http://weibo.com/xingrz)
 * 
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

var mid = {
  BASE62: [
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
    "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", 
    "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", 
    "u", "v", "w", "x", "y", "z",
    "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", 
    "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", 
    "U", "V", "W", "X", "Y", "Z"
  ]
};

/**
 * Base62 decode.
 * 
 * @param {String} str
 * @return {String}
 */
mid.base62Decode = function (str) {
  var num = 0;
  var len = str.length;
  for (var i = 0; i < len; i++) {
    var n = len - i - 1;
    var s = str[i];
    num += mid.BASE62.indexOf(s) * Math.pow(62, n);
  }
  return num;
};

/**
 * Base64 encode.
 * 
 * @param {String} num
 * @return {String}
 */
mid.base62Encode = function (num) {
  var str = '';
  var r = 0;
  while (num !== 0 && str.length < 100) {
    r = num % 62;
    str = mid.BASE62[r] + str;
    num = Math.floor(num / 62);
  }
  return str;
};

/**
 * Conver base62 hash string to number string.
 * 
 * @param {String} hash weibo hash string, e.g.: "wr4mOFqpbO"
 * @return {String} number string, e.g.: "201110410216293360"
 */
mid.decode = function (hash) {
  var id = '';

  for (var end = hash.length; end > 0; end -= 4) {
    var start = end - 4;
    if (start < 0) {
      start = 0;
    }
    var h = hash.substring(start, end);
    var n = String(mid.base62Decode(h));
    var padding = 7 - n.length;
    if (padding > 0 && start > 0) {
      // not first group and not 7 length string, must add '0' padding.
      for (; padding > 0; padding--) {
        n = '0' + n;
      }
    }
    id = n + id;
  }
  
  return id;
};

/**
 * Convert number string to base62 hash string.
 * 
 * @param {String} mid weibo mid, e.g.: "201110410216293360"
 * @return {String} hash string, e.g.: "wr4mOFqpbO"
 */
mid.encode = function (id) {
  id = String(id);
  if (!/^\d+$/.test(id)) { 
    return id; 
  }

  var hash = '';
  for (var end = id.length; end > 0; end -= 7) {
    var start = end - 7;
    if (start < 0) {
      start = 0;
    }
    var num = id.substring(start, end);
    var h = mid.base62Encode(num);
    var padding = 4 - h.length;
    if (padding > 0 && start > 0) {
      // not first group and not 4 length string, must add '0' padding.
      for (; padding > 0; padding--) {
        h = '0' + h;
      }
    }
    hash = h + hash;
  }
  
  return hash;
};

module.exports = mid;


mid.py
import sys

# https://gist.github.com/mrluanma/6000424

ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"


def rsplit(s, count):
    f = lambda x: x > 0 and x or 0
    return [s[f(i - count):i] for i in range(len(s), 0, -count)]


def mid2str(mid):
    result = ''
    for i in rsplit(mid, 7):
        str62 = base62_encode(int(i))
        result = str62.zfill(4) + result
    return result.lstrip('0')


def str2mid(input):
    result = ''
    for i in rsplit(input, 4):
        str10 = str(base62_decode(i)).zfill(7)
        result = str10 + result
    return result.lstrip('0')


def base62_encode(num, alphabet=ALPHABET):
    """Encode a number in Base X

    `num`: The number to encode
    `alphabet`: The alphabet to use for encoding
    """
    if (num == 0):
        return alphabet[0]
    arr = []
    base = len(alphabet)
    while num:
        rem = num % base
        num = num // base
        arr.append(alphabet[rem])
    arr.reverse()
    return ''.join(arr)


def base62_decode(string, alphabet=ALPHABET):
    """Decode a Base X encoded string into the number

    Arguments:
    - `string`: The encoded string
    - `alphabet`: The alphabet to use for encoding
    """
    base = len(alphabet)
    strlen = len(string)
    num = 0

    idx = 0
    for char in string:
        power = (strlen - (idx + 1))
        num += alphabet.index(char) * (base ** power)
        idx += 1
    return num


def test():
    print mid2str('1')
    print mid2str('1024012131')
    print str2mid('A09LvzUOy')
    assert mid2str('231101124784859058') == 'Bh0tgako3U'
    assert str2mid('Bh0tgako3U') == '231101124784859058'

    assert mid2str('3491273850170657') == 'yCirT0Iox'
    assert str2mid('yCirT0Iox') == '3491273850170657'


if __name__ == '__main__':
    test()